[docs] Adding a new model documentation (#322)
facebookresearch/mmf: website/docs/tutorials/concat_bert.md

---
id: concat_bert
title: 'Tutorial: Adding a model - Concat BERT'
sidebar_label: Adding a model - Concat BERT
---

> In this tutorial we will go through the step-by-step process of creating a new model using MMF. In this case, we will create a fusion model for the [Hateful Memes dataset](https://github.com/facebookresearch/mmf/tree/master/projects/hateful_memes).

Reviewer: ```suggestion
In this tutorial, we will go through the step-by-step process of creating a new model using MMF. In this case, we will create a fusion model which we will train on the [Hateful Memes dataset](https://github.com/facebookresearch/mmf/tree/master/projects/hateful_memes).
```

> The fusion model that we will create concatenates embeddings from a text encoder and an image encoder and passes them through a two-layer classifier. MMF provides standard image and text encoders out of the box. For image encoder, we will use ResNet152 image encoder and for text encoder, we will use BERT Base Encoder.

Reviewer: ```suggestion
The fusion model that we will create concatenates embeddings from a text encoder and an image encoder and passes them through a two-layer classifier. MMF provides standard image and text encoders out of the box. For the image encoder, we will use the ResNet152 image encoder and for the text encoder, we will use the BERT-Base Encoder.
```

> ## Installation and Preparing the dataset

Reviewer: Change this section to prerequisites?

> Follow the prerequisites for installation and dataset [here](https://github.com/facebookresearch/mmf/tree/master/projects/hateful_memes#prerequisites).

Reviewer: link the installation docs directly?

## Using MMF to build the model

> We will start building our model `ConcatBERT` using the various building blocks available in MMF. Helper builder methods like  `build_image_encoder` for building image encoders, `build_text_encoder` for building text encoders, `build_classifier_layer` for classifier layers etc take configurable params which are defined in the config we will create in the next section.

Reviewer: ```suggestion
We will start building our model `ConcatBERT` using the various building blocks available in MMF. The helper builder methods like  `build_image_encoder` for building image encoders, `build_text_encoder` for building text encoders, `build_classifier_layer` for classifier layers etc take configurable params which are defined in the config we will create in the next section.

Follow the code and read through the comments to understand the process.
```

        # Similarly, image input will be in "image" key
        image = sample_list["image"]

        # Get the text and image features from the encoders
        text_features = self.language_module(text)[1]
        image_features = self.vision_module(image)

        # Flatten the embeddings before concatenation
        image_features = torch.flatten(image_features, start_dim=1)
        text_features = torch.flatten(text_features, start_dim=1)

        # Concatenate the features returned from two modality encoders
        combined = torch.cat([text_features, image_features], dim=1)

        # Pass final tensor to classifier to get scores
        logits = self.classifier(combined)

        # For loss calculations (automatically done by MMF
        # as per the loss defined in the config),
        # we need to return a dict with "scores" key as logits
        output = {"scores": logits}

        # MMF will automatically calculate loss
        return output

```

The model’s forward method takes a `SampleList` and outputs a dict containing the logit scores predicted by the model. Different losses and metrics can be calculated on the scores output.

> We define two configs needed for our experiments: (i) model config for model defaults (ii) experiment config. Model config provides the defaults for model’s hyperparameters and experiment config defines and overrides the defaults needed for our particular experiment

Reviewer: ```suggestion
We define two configs needed for our experiments: (i) a model config for model's defaults and (ii) an experiment config. The model config provides the defaults for model’s hyperparameters and the experiment config defines and overrides the defaults needed for our particular experiment such as optimizer type, learning rate, maximum updates and batch size.
```

## Model Config

We will now create the model config file with the params we used while creating the model and store the config in `configs/models/concat_bert/defaults.yaml`.

```yaml

model_config:
  concat_bert:

>     # Either pretraining or classification

Reviewer: This seems misplaced.

        num_hidden_layers: 12
        num_attention_heads: 12
        output_attentions: false
        output_hidden_states: false

    classifier:
      type: mlp
      params:
        # 2048 + 768 in case of features
        # Modal_Dim * Number of embeddings + Text Dim
        in_dim: 2816
        out_dim: 2
        hidden_dim: 768
        num_layers: 2

```

## Experiment Config

In the next step, we will create the experiment config which will tell MMF which dataset, optimizer, scheduler, metrics for evalauation to use. We will save this config in `configs/experiments/concat_bert/defaults.yaml`:

```yaml

includes:
- configs/datasets/hateful_memes/bert.yaml

model_config:
  trying:
    classifier:
      type: mlp
      params:
        num_layers: 2
    losses:
    - type: cross_entropy

scheduler:
  type: warmup_linear
  params:
    num_warmup_steps: 2000
    num_training_steps: ${training.max_updates}

optimizer:
  type: adam_w
  params:
    lr: 5e-5
    eps: 1e-8

evaluation:
  metrics:
  - accuracy
  - binary_f1
  - roc_auc

training:
  batch_size: 64
  lr_scheduler: true
  max_updates: 22000
  early_stop:
    criteria: hateful_memes/roc_auc
    minimize: false

```

> We include the `bert.yaml` config in this as we want to use BERT tokenizer for preprocessing our language data. With both the configs ready we are ready to launch training and evaluation using our model on Hateful Memes dataset. You can read more about the MMF’s configuration system [here](https://mmf.sh/docs/notes/configuration).

Reviewer: ```suggestion
We include the `bert.yaml` config in this as we want to use BERT tokenizer for preprocessing our language data. With both the configs ready we are ready to launch training and evaluation using our model on the Hateful Memes dataset. You can read more about the MMF’s configuration system [here](https://mmf.sh/docs/notes/configuration).
```

## Training and Evaluation

Now we are ready to train and evaluate our model with the experiment config we created in previous step.

```bash
mmf_run config="configs/experiments/concat_bert/defaults.yaml" \
    model=concat_bert \
    dataset=hateful_memes \
    run_type=train_val
```

> When training ends it will save a final model `concat_bert_final.pth` in the experiment folder under `./save` directory. The command will also generate validation scores after the training gets over.

Reviewer: Add link to general checkpointing tutorial here about prediction and evaluation.
